src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/TempPECompilerManager.cs

```diff
@@ -0,0 +1,372 @@
+        {
+            if (initialized)
+            {
+                // This will stop our blocks taking any more input
+                _inputsActionBlock?.Complete();
+                _fileWatcherActionBlock?.Complete();
+
+                if (_inputsActionBlock != null)
+                {
+                    await Task.WhenAll(_inputsActionBlock.Completion, _fileWatcherActionBlock!.Completion);
+                }
+
+                // By waiting for completion we know that the following dispose will cancel any pending compilations, and there won't be any more
+                _scheduler.Dispose();
+                _disposables.Dispose();
+            }
+        }
+
+        private async Task ProcessCompileQueue(CancellationToken token)
+        {
+            await _activeWorkspaceProjectContextHost.OpenContextForWriteAsync(async accessor =>
+            {
+                // Grab the next file to compile off the queue
+                (string fileName, bool ignoreFileWriteTime) = _filesToCompile.FirstOrDefault();
+                while (fileName != null)
+                {
+                    if (IsDisposing || IsDisposed)
+                    {
+                        return;
+                    }
+
+                    token.ThrowIfCancellationRequested();
+
+                    // Remove the file from our todo list. If it wasn't there (because it was removed as a design time input while we were busy) we don't need to compile it
+                    bool wasInQueue = ThreadingTools.ApplyChangeOptimistically(ref _filesToCompile, fileName, (s, f) => s.Remove(f));
+
+                    string outputFileName = await GetOutputFileName(fileName);
+                    if (wasInQueue)
+                    {
+                        await CompileDesignTimeInput(accessor.Context, fileName, outputFileName, ignoreFileWriteTime, token);
+                    }
+
+                    // Grab another file off the queue
+                    (fileName, ignoreFileWriteTime) = _filesToCompile.FirstOrDefault();
+                }
+            });
+        }
+
+        /// <summary>
+        /// Gets the XML that describes a TempPE DLL, including building it if necessary
+        /// </summary>
+        /// <param name="fileName"></param>
+        /// <returns></returns>
+        public async Task<string> GetDesignTimeInputXML(string fileName)
+        {
+            // Make sure we're not being asked to compile a random file
+            if (!_designTimeInputs.Contains(fileName, StringComparers.Paths))
+            {
+                throw new ArgumentException("Can only get XML snippets for design time inputs", nameof(fileName));
+            }
+
+            // Remove the file from our todo list, in case it was in there.
+            ThreadingTools.ApplyChangeOptimistically(ref _filesToCompile, fileName, (s, f) => s.Remove(f));
+
+            string outputFileName = await GetOutputFileName(fileName);
```

Comment:
This call will deadlock (in `await _receivedProjectRuleSource.Task;`) if the following is true:

1. This method is entered on the UI thread
2. The data sources for ProcessDataflowChanges need the UI thread to finish. 

In that case, the UI thread will not let step 2. proceed as it doesn't know that it needs to let the await on the UI thread to proceed.